src/Storages/FileLog/StorageFileLog.cpp

```diff
@@ -0,0 +1,746 @@
+    unsigned /* num_streams */)
+{
+    std::lock_guard<std::mutex> lock(status_mutex);
+
+    updateFileInfos();
+
+    /// No files to parse
+    if (file_infos.file_names.empty())
+    {
+        return Pipe{};
+    }
+
+    auto modified_context = Context::createCopy(local_context);
+
+    auto max_streams_number = std::min<UInt64>(filelog_settings->filelog_max_threads, file_infos.file_names.size());
+
+    Pipes pipes;
+    pipes.reserve(max_streams_number);
+    for (size_t stream_number = 0; stream_number < max_streams_number; ++stream_number)
+    {
+        Pipe pipe(std::make_shared<FileLogSource>(
+            *this, metadata_snapshot, modified_context, getMaxBlockSize(), getPollTimeoutMillisecond(), stream_number, max_streams_number));
+
+        auto convert_actions_dag = ActionsDAG::makeConvertingActions(
+            pipe.getHeader().getColumnsWithTypeAndName(),
+            metadata_snapshot->getSampleBlockForColumns(column_names, getVirtuals(), getStorageID()).getColumnsWithTypeAndName(),
+            ActionsDAG::MatchColumnsMode::Name);
+
+        auto actions = std::make_shared<ExpressionActions>(
+            convert_actions_dag, ExpressionActionsSettings::fromContext(getContext(), CompileExpressions::yes));
+
+        pipe.addSimpleTransform([&](const Block & stream_header) { return std::make_shared<ExpressionTransform>(stream_header, actions); });
+        pipes.emplace_back(std::move(pipe));
+    }
+
+    return Pipe::unitePipes(std::move(pipes));
+}
+
+void StorageFileLog::drop()
+{
+    if (std::filesystem::exists(root_meta_path))
+        std::filesystem::remove_all(root_meta_path);
+}
+
+void StorageFileLog::startup()
+{
+    if (task)
+    {
+        task->holder->activateAndSchedule();
+    }
+}
+
+void StorageFileLog::shutdown()
+{
+    if (task)
+    {
+        task->stream_cancelled = true;
+
+        LOG_TRACE(log, "Waiting for cleanup");
+        task->holder->deactivate();
+    }
+    closeFilesAndStoreMeta();
+}
+
+void StorageFileLog::openFilesAndSetPos()
+{
+    for (const auto & file : file_infos.file_names)
+    {
+        auto & file_ctx = file_infos.context_by_name.at(file);
+        if (file_ctx.status != FileStatus::NO_CHANGE)
+        {
+            file_ctx.reader.open(getFullDataPath(file));
+            if (!file_ctx.reader.good())
+            {
+                throw Exception(ErrorCodes::FILE_STREAM_ERROR, "Open file {} failed.", file);
+            }
+
+            file_ctx.reader.seekg(0, file_ctx.reader.end);
+            if (!file_ctx.reader.good())
+            {
+                throw Exception(ErrorCodes::FILE_STREAM_ERROR, "Seekg file {} failed.", file);
+            }
+
+            auto file_end = file_ctx.reader.tellg();
+            if (!file_ctx.reader.good())
+            {
+                throw Exception(ErrorCodes::FILE_STREAM_ERROR, "Tellg file {} failed.", file);
+            }
+
+            auto & meta = file_infos.meta_by_inode.at(file_infos.inode_by_name.at(file));
```

Comment:
lets split this line in two.
Also this is the second time I see this line and I also started thinking that we should use `find` instead of `at` and throw logical error if not found. And this logic can be moved to a static method here.